src/array/dictionary/value_map.rs

```diff
@@ -0,0 +1,126 @@
+use std::borrow::Borrow;
+use std::fmt::{self, Debug};
+use std::hash::{BuildHasher, Hash, Hasher};
+
+use hashbrown::hash_map::RawEntryMut;
+use hashbrown::HashMap;
+
+use crate::array::Array;
+use crate::{
+    array::indexable::{AsIndexed, Indexable},
+    array::MutableArray,
+    datatypes::DataType,
+    error::{Error, Result},
+};
+
+use super::DictionaryKey;
+
+pub struct ValueMap<K: DictionaryKey, M: MutableArray> {
+    values: M,
+    map: HashMap<K, ()>, // NB: *only* use insert_hashed_nocheck() and no other hashmap API
+}
+
+impl<K: DictionaryKey, M: MutableArray> ValueMap<K, M> {
+    pub fn try_empty(values: M) -> Result<Self> {
+        if !values.is_empty() {
+            return Err(Error::InvalidArgumentError(
+                "initializing value map with non-empty values array".into(),
+            ));
+        }
+        Ok(Self {
+            values,
+            map: HashMap::default(),
+        })
+    }
+
+    pub fn from_values(values: M) -> Result<Self>
+    where
+        M: Indexable,
+        M::Type: Eq + Hash,
+    {
+        let mut map = HashMap::with_capacity(values.len());
+        for index in 0..values.len() {
+            let key = K::try_from(index).map_err(|_| Error::Overflow)?;
+            // safety: we only iterate within bounds
+            let value = unsafe { values.value_unchecked_at(index) };
+            let mut hasher = map.hasher().build_hasher();
+            value.borrow().hash(&mut hasher);
+            let hash = hasher.finish();
+            match map.raw_entry_mut().from_hash(hash, |_| true) {
+                RawEntryMut::Occupied(_) => {
+                    return Err(Error::InvalidArgumentError(
```

Comment:
Don't we want to allow it and just use the first occurring? No strong opinion on my part, but it might be handy.

Reply:
No strong preference either! Can change if you'd prefer it this way. I think my point was, there's cases where you really want to ensure that values are unique (e.g. writing chunks).

But it's actually not the best way of doing it anyway for those use cases... I just realized one part that's missing is a method allowing to clone value map and reset the keys (without taking ownership), perhaps something like `to_empty(&self)`.

Reply:
I prefer to throw an error if the values are not unique.

Reply:
1 vote versus zero, leave it as ensuring it's unique then?

Unless there's anything else, should be lgtm I guess.